streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -208,31 +205,57 @@ public synchronized void start() {
 
     /**
      * Shutdown this stream instance by signaling all the threads to stop,
-     * and then wait for them to join.
+     * and then wait for them to join. Uses a default timeout of 30 seconds
+     */
+    public void close() {
+        close(DEFAULT_CLOSE_TIMEOUT, TimeUnit.SECONDS);
+    }
+
+    /**
+     * Shutdown this stream instance by signaling all the threads to stop,
+     * and then wait up to the timeout for the threads to join.
      *
-     * @throws IllegalStateException if process has not started yet
+     * @param timeout   how long to wait for {@link StreamThread}s to shutdown
+     * @param timeUnit  unit of time used for timeout
      */
-    public synchronized void close() {
+    public synchronized void close(final long timeout, final TimeUnit timeUnit) {
         log.debug("Stopping Kafka Stream process");
-
-        if (state == RUNNING) {
-            // signal the threads to stop and wait
-            for (final StreamThread thread : threads)
-                thread.close();
-
-            for (final StreamThread thread : threads) {
-                try {
-                    thread.join();
-                } catch (final InterruptedException ex) {
-                    Thread.interrupted();
-                }
+        if (state == StreamsState.running) {
+            // save the current thread so that if it is a stream thread
+            // we don't attempt to join it and cause a deadlock
+            final Thread shutdown = new Thread(new Runnable() {
+                @Override
+                public void run() {
+                        // signal the threads to stop and wait
+                        for (final StreamThread thread : threads) {
+                            thread.close();
```

Comment:
If `streams.close()` is called by one of its own stream thread, then do we still have a deadlock, where the stream thread wait for the shutdown stream to join and the shutdown thread wait for the stream thread to join?